WordPress/Classes/Utility/CoreDataHelper.swift

```diff
@@ -426,3 +433,81 @@ extension CoreDataStack {
         }
     }
 }
+
+/// `CoreDataObjectIdentifier` is a strongly typed `NSManagedObject`.
+///
+/// By using this strongly typed `NSManagedObject`, we can declare APIs that're bound to specific model types, thus
+/// preventing an incorrect object id being used as argument.
+///
+/// Take the following function as an example,
+/// ```
+/// func getPost(fromBlogID blogID: NSManagedObjectID)
+/// ```
+///
+/// We can call it with any `NSManagedObjectID`: `getPost(fromBlogID: themeObjectID)`. This usage is obviously
+/// incorrect, but we won't be able to catch this error until something goes wrong at rumtime.
+///
+/// However, we can change the declaration to
+/// ```
+/// func getPost(fromBlogID: CoreDataObjectIdentifier<Blog>)
+/// ```
+///
+/// Now the type `Blog` is built into the function signature, and Swift compiler would report an error if we call it using
+/// `getPost(fromBlogID: themeObjectID)`.
+struct CoreDataObjectIdentifier<Model: NSManagedObject> {
```

Comment:
An alternative is using [swift-tagged](https://github.com/pointfreeco/swift-tagged) and declaring the strongly typed id types like:

```swift
typealias BlogObjectID = Tagged<Blog, NSManangedObjectID>
typealias PostObjectID = Tagged<Post, NSManangedObjectID>
// ...
```

and they can be used like:

```swift
func getPost(from blogID: BlogObjectID) async -> PostObjectID
```

Reply:
I'd be interested to see `swift-tagged` in action as it could bring benefits to other parts of the codebase.

Part of me is tempted to say, introduce it only when you need it. But in this case `swift-tagged` would act as an enabler type that we could build on top of. Also, I wonder how costly it would be to modify this code if we decided to replace it with the library in the future 🤔

Reply:
> Also, I wonder how costly it would be to modify this code if we decided to replace it with the library in the future 🤔

Quite straight forward actually. Only need to rename a few types: https://github.com/wordpress-mobile/WordPress-iOS/pull/21237

Reply:
I don't think we should use `swift-tagged` for this use case. The `Tagged` type has a public initializer, which means we can't validate that the given object id is really the declared model type. In another word, it's possible to create a `let postID = CoreDataIdentifier<Post>(rawValue: account.objectID)`. I don't think we should allow that...